src/oatpp/core/provider/Pool.hpp

```diff
@@ -210,79 +237,64 @@ class PoolTemplate : public oatpp::base::Countable, public async::CoroutineWaitL
-
-    {
-
-      std::unique_lock<std::mutex> guard(m_lock);
+    auto finishedPredicate = [&_this]() { return !_this->m_running || !_this->m_bench.empty() || _this->m_counter < _this->m_maxResources; };
+    if (!_this->m_condition.wait_for(guard, timeout, std::move(finishedPredicate))) {
+      return nullptr;
+    }
 
-      while (m_running && m_bench.size() == 0 && m_counter >= m_maxResources ) {
-        m_condition.wait(guard);
-      }
+    return get(_this, std::move(guard));
 
-      if(!m_running) {
-        return nullptr;
-      }
+  }
 
-      if (m_bench.size() > 0) {
-        auto record = m_bench.front();
-        m_bench.pop_front();
-        return std::make_shared<AcquisitionProxyImpl>(record.resource, _this);
-      } else {
-        ++ m_counter;
-      }
+  static std::shared_ptr<TResource> get(const std::shared_ptr<PoolTemplate>& _this) {
+    
+    std::unique_lock<std::mutex> guard(_this->m_lock);
 
+    while (_this->m_running && _this->m_bench.empty() && _this->m_counter >= _this->m_maxResources ) {
+        _this->m_condition.wait(guard);
     }
 
-    try {
-      return std::make_shared<AcquisitionProxyImpl>(m_provider->get(), _this);
-    } catch (...) {
-      std::lock_guard<std::mutex> guard(m_lock);
-      -- m_counter;
-      return nullptr;
-    }
+    return get(_this, std::move(guard));
 
   }
 
-  async::CoroutineStarterForResult<const std::shared_ptr<TResource>&> getAsync(const std::shared_ptr<PoolTemplate>& _this) {
+  static async::CoroutineStarterForResult<const std::shared_ptr<TResource>&> getAsync(const std::shared_ptr<PoolTemplate>& _this, const std::chrono::duration<v_int64, std::micro>& timeout) {
 
     class GetCoroutine : public oatpp::async::CoroutineWithResult<GetCoroutine, const std::shared_ptr<TResource>&> {
     private:
       std::shared_ptr<PoolTemplate> m_pool;
-      std::shared_ptr<Provider<TResource>> m_provider;
+
+      std::chrono::steady_clock::time_point m_startTime{std::chrono::steady_clock::now()};
+      std::chrono::duration<v_int64, std::micro> m_timeout;
     public:
 
-      GetCoroutine(const std::shared_ptr<PoolTemplate>& pool, const std::shared_ptr<Provider<TResource>>& provider)
+      GetCoroutine(const std::shared_ptr<PoolTemplate>& pool, const std::chrono::duration<v_int64, std::micro>& timeout)
         : m_pool(pool)
-        , m_provider(provider)
+        , m_timeout(timeout)
       {}
 
+      bool timedout() const noexcept {
+        return m_timeout != std::chrono::microseconds::zero() && m_timeout < (std::chrono::steady_clock::now() - m_startTime);
+      }
+
       async::Action act() override {
 
         {
           /* Careful!!! Using non-async lock */
-          std::unique_lock<std::mutex> guard(m_pool->m_lock);
+          std::unique_lock<std::mutex> guard(m_pool->m_lock, std::try_to_lock);
+
+          if (!guard.owns_lock()) {
+            return timedout() ? this->_return(nullptr) : this->waitRepeat(std::chrono::milliseconds(100));
```

Comment:
I had a look at `CoroutineWaitList` and the `Action` Union-Member `Data` and was wondering how we would pass the timeout. 
Would it be ok to create a new struct which incapsulates the `CoroutineWaitList*` and a `std::duration::steady_clock::timep_point` to represent the time the timeout would be exceeded?

Reply:
Hey @MHaselmaier ,

Sorry, just found your comment - I guess I missed it during the previous review.

> I had a look at CoroutineWaitList and the Action Union-Member Data and was wondering how we would pass the timeout.

We want to keep `Action` as small as possible and for this reason, we don't want to put big structures in the union. 
However, since `union Data`  already has a `FunctionPtr` member which takes 16 bytes - we have plenty of place to put there a pointer to WaitList + timeout.

```cpp
  struct WaitListWithTimeout {
    CoroutineWaitList* waitList;
    v_uint64 timeout; // or std::chrono::steady_clock::time_point timeout
  };

  union Data {
    FunctionPtr fptr;
    AbstractCoroutine* coroutine;
    Error* error;
    IOData ioData;
    v_int64 timePointMicroseconds;
    WaitListWithTimeout waitListWithTimeout;
  };
```